homeassistant/components/bryant_evolution/climate.py

```diff
@@ -0,0 +1,241 @@
+"""Support for Bryant Evolution HVAC systems."""
+
+from datetime import timedelta
+import logging
+from typing import Any
+
+from evolutionhttp import BryantEvolutionClient
+
+from homeassistant.components.climate import (
+    ClimateEntity,
+    ClimateEntityFeature,
+    HVACAction,
+    HVACMode,
+)
+from homeassistant.const import CONF_HOST, UnitOfTemperature
+from homeassistant.core import HomeAssistant
+from homeassistant.exceptions import HomeAssistantError
+from homeassistant.helpers.device_registry import DeviceInfo
+from homeassistant.helpers.entity_platform import AddEntitiesCallback
+
+from . import BryantEvolutionConfigEntry
+from .const import CONF_SYSTEM_ID, CONF_ZONE_ID, DOMAIN
+
+_LOGGER = logging.getLogger(__name__)
+
+
+SCAN_INTERVAL = timedelta(seconds=60)
+
+
+async def async_setup_entry(
+    hass: HomeAssistant,
+    config_entry: BryantEvolutionConfigEntry,
+    async_add_entities: AddEntitiesCallback,
+) -> None:
+    """Set up a config entry."""
+    client = config_entry.runtime_data
+    climate = BryantEvolutionClimate(
+        config_entry.data[CONF_HOST],
+        config_entry.data[CONF_SYSTEM_ID],
+        config_entry.data[CONF_ZONE_ID],
+        client,
+    )
+    async_add_entities([climate], update_before_add=True)
+
+
+class BryantEvolutionClimate(ClimateEntity):
+    """ClimateEntity for Bryant Evolution HVAC systems.
+
+    Design note: this class updates using polling. However, polling
+    is very slow (~1500 ms / parameter). To improve the user
+    experience on updates, we also locally update this instance and
+    call async_write_ha_state as well.
+    """
+
+    _attr_has_entity_name = True
+
+    def __init__(
+        self: ClimateEntity,
+        host: str,
+        system_id: int,
+        zone_id: int,
+        client: BryantEvolutionClient,
+    ) -> None:
+        """Initialize an entity from parts."""
+        self._client = client
+        self._attr_name = None
+        self._attr_temperature_unit = UnitOfTemperature.FAHRENHEIT
+        self._attr_supported_features = (
+            ClimateEntityFeature.TARGET_TEMPERATURE
+            | ClimateEntityFeature.TARGET_TEMPERATURE_RANGE
+            | ClimateEntityFeature.FAN_MODE
+            | ClimateEntityFeature.TURN_ON
+            | ClimateEntityFeature.TURN_OFF
+        )
+        self._attr_hvac_modes = [
+            HVACMode.HEAT,
+            HVACMode.COOL,
+            HVACMode.HEAT_COOL,
+            HVACMode.OFF,
+        ]
+        self._attr_fan_modes = ["AUTO", "LOW", "MED", "HIGH"]
+        self._enable_turn_on_off_backwards_compatibility = False
+        self._attr_unique_id = f"bryant_evolution_{host}_{system_id}_{zone_id}"
```

Comment:
Also, the `host` is not unique as it can change. Do we have something truly unique from the device?

Reply:
Unfortunately, we do not -- the protocol does not support a system-level ID for the HVAC equipment. I could add a unique ID to the server that exposes the Bryant SAM over HTTP (which I maintain), then provide a way to flow that to HA, but I don't know how much that would add over just keying on host and (say) using DNS?

Reply:
Wait, can you elaborate on that? Is this an integration for the device or fit the service that integrates with the device

Reply:
There are three components here, plus HA:

[Furnace (1)] <-- RS-485 --> [SAM (2)] <--RS-232 -->[HTTP server (3b)] <--HTTP--> [HomeAssistant (3a)] 

1) A Bryant-manufactured HVAC system (furnace, air conditioner, etc). This internally uses a proprietary protocol over RS-485 to communicate with its thermostats and HVAC components.

2) A Bryant-manufactured "System Access Module" (SAM). This is hardware that exposes an ASCII interface over RS-232. The SAM converts between the ASCII commands and the proprietary protocol. The SAM protocol reference is [here](https://www.shareddocs.com/hvac/docs/1009/Public/06/APSAM01-03.pdf) if you're curious.

3) An HTTP client (3a, Python, linked into HA) and server (3b, running on a device with a serial port to which the SAM is connected, such as a Pi Zero W) that exposes the SAM over the network. This is https://github.com/danielsmyers/evolutionhttp. It more or less just exposes the serial port over HTTP -- it accepts and returns the ASCII commands from the SAM protocol. It also does some queue management to reorder writes in front of reads, because the SAM is quite slow, and I wanted to prioritize interactive traffic.

So, I'd say this is an integration for Bryant HVAC systems. It requires that a System Access Module be a part of the system. To allow HA to talk to the SAM, there's a bit of middleware I wrote.

Does that help?

Reply:
can there be more than one device per system/zone combination? Otherwise can we maybe use that as identifier?

Reply:
The thing that bothers me is that we don't directly integrate with the device and have to use another project to actually make it work. That doesn't fit in core in my opinion.

Reply:
Modified to connect to the device directly over a serial port (note we now use a BryantEvolutionLocalClient, not a BryantEvolutionClient).

Reply:
I'm still unclear what we should use as an identifier, however -- there's no programmatic way to retrieve a unique ID from the SAM.

In theory, one could have two SAMs connected to HA. Each of them would have (say) a "system 1, zone 1". So, a system/zone pairing is not unique. 

WDYT?

Reply:
Continuing to think about unique IDs...

We now have an architecture that looks like this:

HVAC <--RS-485--> SAM <---RS-232 --> Home Assistant

Recall:

1. The SAM is a hardware device manufactured by the same company as the HVAC equipment. It's the thing HA talks to; it internally does an ASCII-to-proprietary protocol translation.
2. A single SAM can expose 1 or 2 HVAC systems (say, an upstairs and a downstairs furnace). They are always referred to as "1" and "2".
3. A single system can have 1 to 8 zones. They are always named 1-8.
4. Climate-like operations (e.g., get current temperature, set target temperature, set fan speed) apply to (system, zone) pairs.
5. As such, we expose entities in HA that map to (system, zone) tuples.

The entities that we expose to Home Assistant need unique IDs. Relevant facts:

1. There is nothing in the protocol exposed by the SAM that allows reading a serial number or any kind of unique ID whatsoever for the SAM itself, the HVAC systems, or their zones.
2. There is a serial number physically printed on the SAM itself. It's on a circuit board, but the cover removes easily.
3. The mapping the SAM maintains from system and zone numbers to actual physical HVAC equipment is a function of how it is wired to the equipment (system numbers) and how the equipment is programmed (zone numbers). I think we can assume that this mapping is effectively stable forever, such that we can ask the user to redo the integration setup if they change any of it.

So, I think we only have two options, based on reading the [docs](https://developers.home-assistant.io/docs/entity_registry_index/)

1. Use the tuple (serial number printed on the SAM, system id, zone id), assuming that system and zone numbers are stable (point 3 above). Presumably we need to add a form field in the config flow to have the user supply the serial number. This is a bit of irritating friction for the user, but it's not *that* bad.

2. Use the config entry ID as the unique ID (last resort option). 

I'm leaning to (1).

WDYT?

Reply:
I'm leaning towards 2

Reply:
done